src/i_input.c

```diff
@@ -782,31 +899,67 @@ void I_HandleKeyboardEvent(SDL_Event *sdlevent)
     {
         case SDL_KEYDOWN:
             event.type = ev_keydown;
-            event.data1.i = TranslateKey(sdlevent->key.keysym.scancode);
+            event.data1.i = TranslateKey(&sdlevent->key.keysym);
+            event.data2.i = GetLocalizedKey(&sdlevent->key.keysym);
 
             if (event.data1.i != 0)
             {
                 D_PostEvent(&event);
             }
+
+            if (text_input_enabled && vanilla_keyboard_mapping)
+            {
+                event.type = ev_text;
+                event.data1.i = GetTypedChar(&sdlevent->key.keysym);
+
+                if (event.data1.i != 0)
+                {
+                    D_PostEvent(&event);
+                }
+            }
             break;
 
         case SDL_KEYUP:
             event.type = ev_keyup;
-            event.data1.i = TranslateKey(sdlevent->key.keysym.scancode);
+            event.data1.i = TranslateKey(&sdlevent->key.keysym);
 
-            // data2/data3 are initialized to zero for ev_keyup.
-            // For ev_keydown it's the shifted Unicode character
-            // that was typed, but if something wants to detect
-            // key releases it should do so based on data1
+            // data2 is initialized to zero for ev_keyup. For ev_keydown it's
+            // the shifted Unicode character that was typed, but if something
+            // wants to detect key releases it should do so based on data1
             // (key ID), not the printable char.
 
+            event.data2.i = 0;
+
             if (event.data1.i != 0)
             {
                 D_PostEvent(&event);
             }
             break;
 
+        case SDL_TEXTINPUT:
+            if (text_input_enabled
```

Comment:
I wonder if this requires an additional `&& !vanilla_keyboard_mapping` check here? Else, a single key press, which triggers both an `SDL_KEYDOWN` event and a subsequent `SDL_TEXTINPUT` event in SDL, will post an `ev_text` event in the `case SDL_KEYDOWN` switch branch and another `ev_text` event here in the `case SDL_TEXTINPUT` switch branch.

Reply:
We don't run `SDL_StartTextInput()` in case `vanilla_keyboard_mapping == true`, so SDL doesn't fire `SDL_TEXTINPUT` events.

Reply:
Done in 97f906c